crates/bevy_audio/src/audio_input.rs

```diff
@@ -0,0 +1,180 @@
+use std::sync::mpsc::{channel, Receiver};
+
+use bevy_ecs::{prelude::*, system::Command};
+use bevy_utils::{
+    tracing::{trace, warn},
+    Duration,
+};
+use rodio::cpal::{
+    self,
+    traits::{DeviceTrait, HostTrait, StreamTrait},
+    InputCallbackInfo, PauseStreamError, PlayStreamError, Stream, StreamConfig,
+};
+
+/// Used internally to retrieve audio on the current "audio input device". When this is dropped,
+/// the underlying stream and event system are also dropped.
+pub struct AudioInputStream {
+    config: StreamConfig,
+    stream: Stream,
+    receiver: Receiver<AudioInputEvent>,
+}
+
+/// A frame of samples from an audio input source.
+#[derive(Event, Clone)]
+pub struct AudioInputEvent {
+    /// Interleaved samples recorded during this frame of audio input. Please check [`config`](`Self::config`)
+    /// for channel information and sample rate.
+    pub samples: Vec<f32>,
+    /// Information about this sample frame as provided by [`cpal`].
+    pub info: InputCallbackInfo,
+    /// Configuration used for the input device which recorded this sample frame.
+    pub config: StreamConfig,
+}
+
+impl AudioInputStream {
+    /// Try to create a default audio input stream. Returns [`None`] if a default input stream
+    /// cannot be created.
+    pub fn try_default() -> Option<Self> {
+        let host = cpal::default_host();
+        let input = host.default_input_device()?;
+        let config = input.default_input_config().ok()?.config();
+
+        trace!("Got Audio Input Device: {:?}: {:?}", input.name(), config);
+
+        let (tx, receiver) = channel();
+
+        // Create a clone of the configuration for the callback
+        let config_clone = config.clone();
+
+        let stream = input
+            .build_input_stream(
+                &config,
+                move |data: &[f32], info: &InputCallbackInfo| {
+                    let event = AudioInputEvent {
+                        samples: data.to_vec(),
+                        info: info.clone(),
+                        config: config_clone.clone(),
+                    };
+
+                    if let Err(error) = tx.send(event) {
+                        warn!("Audio Input Error: {}", error);
+                    };
```

Comment:
Allocations in audio threads are discouraged. To learn more about why, I recommend this article: http://www.rossbencina.com/code/real-time-audio-programming-101-time-waits-for-nothing

In short, audio threads are supposed to never be paused while the data callback is executing. Because any system call (such as asking the OS for memory) can cause the OS to pause the thread, the callback might run for longer than the duration of an audio buffer, and the audio will start to glitch due to "buffer underrun/underflow", that is audio input data is being fed into the callback function faster than it can process it.

The callback here (potentially) asks for memory in two places: `data.to_vec()` and `tx.send(event)`, the former at each invocation of the callback, and the latter in cases where the channel is filling up faster than it is emptying.

In general, audio programming needs to be real-time safe, that is, wait-free (that is, progress is guaranteed) and lock-free or non-blocking (threads must not be paused). Only structures that have both these characteristics should be used here, and in this case, the solution is a circular buffer (eg. the `rtrb` crate), that is preallocated once and offers both writes and reads that are both wait-free and lock-free.

For this kind of application, the circular buffer can directly hold audio data, without the need for an extra `AudioInputEvent` struct, as the `info` and `config` are not going to change for the duration of the input stream, and the contiguous layouts will help with cache hit rates. For example with the `rtrb` crate, the producer and consumer can both write (resp. read) chunks of audio data directly into the buffer at once.

Using a circular buffer means that if it is full, the oldest data is dropped, which results in glitches as well, but the callback execution time is still bounded, and if the consumer of this audio data is running late, well that's nothing we can do about it and not our problem. 😆 

As for how big the circular buffer should be, a first pass could be set at a multiple of the input data size (something like 3 or 5 times the buffer, here we could define `let circ_buffer_size = 5 * config.buffer_size`), however for safety we could envison having a entire second of audio buffer available, so `let circ_buffer_size = config.sample_rate as usize`. Measurements on how much space is actually used could be carried out on different machines, and a 99.9% percentile could be used as the buffer size.

Reply:
Thanks for the detailed write-up on this! I really appreciate the amount of work this represents, and I'm deeply sorry I didn't respond to you sooner. I wanted to make sure I had a good chunk of free time to really take this all on-board, and I finally had some on the commute between some family functions this weekend.

What I've done to address this issue is two-fold:

1. Switched to using bounded `sync_channel`'s instead of unbounded `channel`'s.
2. Added a second channel for sending pre-allocated sample buffers between the audio input callback (the hot-path) and the rest of Bevy.

Since the `sync_channel` is bounded, it is backed by a fixed (and pre-allocated) array, so sending events will not block, and not allocate. To handle audio buffer vectors, I've created a second channel which `Bevy` can use to send pre-allocated buffers back to the audio system. This could be thought of as recycling the existing buffers, but to allow the event lifetime to be unrestricted, I'm just making new vectors within the event system 1:1 with sent events. To handle the case where my predicted buffer size is inadequate, the audio hot path _can_ allocate to extend the vector it is provided, and the new length will be used for further pre-allocated buffers. As long as the audio system sends buffers of constant size (which `cpal` is supposed to do), they _shouldn't_ allocate in the hot path.

While I agree a ring buffer would be the more direct solution to this problem, this channel based alternative has the same end result, with the added benefit of allowing re-allocation if it has to happen, and only relying on standard library primitives.

I personally haven't performance tested this implementation, but it does work for the provided example, and appears to be low enough latency for my intended use case (voice-chat).